legion/engine/llri-vk/detail/command_group.cpp

```diff
@@ -0,0 +1,130 @@
+
+        m_cmdLists.emplace(output);
+
+        *cmdList = output;
+        return result::Success;
+    }
+
+    result CommandGroup::impl_allocate(const command_list_alloc_desc& desc, uint8_t count, std::vector<CommandList*>* cmdLists)
+    {
+        VkCommandBufferAllocateInfo allocInfo { VK_STRUCTURE_TYPE_COMMAND_BUFFER_ALLOCATE_INFO, nullptr };
+        allocInfo.commandPool = static_cast<VkCommandPool>(m_ptr);
+        allocInfo.commandBufferCount = count;
+        allocInfo.level = internal::mapCommandListUsage(desc.usage);
+
+        std::vector<VkCommandBuffer> cmdBuffers(count);
+        auto r = static_cast<VolkDeviceTable*>(m_deviceFunctionTable)->vkAllocateCommandBuffers(static_cast<VkDevice>(m_device->m_ptr), &allocInfo, cmdBuffers.data());
+        if (r != VK_SUCCESS)
+            return internal::mapVkResult(r);
+
+        for (uint8_t i = 0; i < count; i++)
+        {
+            auto* cmdList = new CommandList();
+            cmdList->m_ptr = cmdBuffers[i];
+            cmdList->m_group = this;
+
+            cmdList->m_deviceHandle = m_device->m_ptr;
+            cmdList->m_deviceFunctionTable = m_deviceFunctionTable;
+
+            cmdList->m_nodeMask = desc.nodeMask;
+            cmdList->m_usage = desc.usage;
+            cmdList->m_state = command_list_state::Empty;
+
+            cmdList->m_validationCallback = m_validationCallback;
+            cmdList->m_validationCallbackMessenger = m_validationCallbackMessenger;
+
+            m_cmdLists.emplace(cmdList);
+            cmdLists->push_back(cmdList);
+        }
+
+        return result::Success;
+    }
+
+    result CommandGroup::impl_free(CommandList* cmdList)
+    {
+        //Free internal pointer
+        auto buffer = static_cast<VkCommandBuffer>(cmdList->m_ptr);
+        static_cast<VolkDeviceTable*>(m_deviceFunctionTable)
+            ->vkFreeCommandBuffers(static_cast<VkDevice>(m_device->m_ptr), static_cast<VkCommandPool>(m_ptr), 1, &buffer);
+
+        //Remove from commandlist list
+        m_cmdLists.erase(cmdList);
+
+        //Delete wrapper
+        delete cmdList;
+        return result::Success;
+    }
+
+    result CommandGroup::impl_free(uint8_t numCommandLists, CommandList** cmdLists)
+    {
+        //Gather vk handles
+        std::vector<VkCommandBuffer> buffers(numCommandLists);
+        for (uint8_t i = 0; i < numCommandLists; i++)
+            buffers[i] = static_cast<VkCommandBuffer>(cmdLists[i]->m_ptr);
+
+        //Free internal pointers
+        static_cast<VolkDeviceTable*>(m_deviceFunctionTable)
+            ->vkFreeCommandBuffers(static_cast<VkDevice>(m_device->m_ptr), static_cast<VkCommandPool>(m_ptr), static_cast<uint32_t>(buffers.size()), buffers.data());
+
+        for (uint8_t i = 0; i < numCommandLists; i++)
+        {
+            //Remove from commandlist list
+            m_cmdLists.erase(cmdLists[i]);
+
+            //Delete wrapper
+            delete cmdLists[i];
+        }
```

Comment:
again here, the use of uint8_t saves no space due to 64bit registers and only costs time due to having to cast numCommandLists to a size_t  in the vector constructor, and `i` being cast to a size_t every iteration